### PR TITLE
Updates to Strategy Construction for TP4

### PR DESCRIPTION
**Pull request by Cole-Greer**
This PR is based off of a recent [dev list post regarding changes to strategy construction](https://lists.apache.org/thread/xv235syvzgl06qcsqr0rrx8boch7ozjp) in response to bytecode removal.

This PR does not contain any changes related to adapting TraversalStrategyProxy to construct custom strategies, the changes here include fixes to allow TinkerPop provided strategies to be constructed and configured remotely, as well as the addition of feature tests for TinkerPop provided strategies.

In general, the feature tests don't have the granularity to verify the strategy is active and functioning correctly. The main intention of these tests is to ensure that the GLVs can use all expected strategies in both `withStrategies()` and `withoutStrategies()` and pass configurations where applicable. For each scenario, I generally tried to use a traversal in which the strategy will have an affect, although in most cases it is not verifiable that the strategy acted as expected. Testing of strategy behaviour is left to the pre-existing unit tests. I would like to extend these tests with scenarios which take advantage of `explain()` once [TINKERPOP-2128](https://issues.apache.org/jira/browse/TINKERPOP-2128) is resolved.

Changes included in this PR:

1. Add missing strategies to groovy's core imports, and the grammar's global strategy registry.
2. Update groovy translator to escape '$' in string literals. Unescaped '$' indicates an embedded value in groovy's GString literals.
3. Update python, .Net, groovy, go, and JS traversalStrategy translators to handle optional `new` in gremlin-lang strategy declaration. eg. `withStrategies(new ReadOnlyStrategy())` as well as `withStrategies(ReadOnlyStrategy())`
4. Change signature of OptionsStrategy in python to be both more pythonic, and bear closer resemblance to gremlin-lang. For example, `OptionsStrategy(options={'myVar': 'myValue'})` has been updated to `OptionsStrategy(myVar='myValue')`
5. Add missing strategies to python's strategies.py
6. Add static `instance()` method to obtain ElementIdStrategy with default configuration.
7. Updated ElementIdStrategy `getConfiguration()` such that `idMaker` is excluded from the config if it is set to the default id supplier. This ensures that configurations which do not override the default idMaker remain serializable.

### CODE REVIEW

**Comment by Cole-Greer** [comment]
VOTE +1

**Comment by xiazcy** [comment]
VOTE +1

**Comment by kenhuuu** [comment]
VOTE +1, pending very small question

File: gremlin-test/src/main/resources/org/apache/tinkerpop/gremlin/test/features/integrated/ConnectiveStrategy.feature [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      g.withStrategies(ConnectiveStrategy).V().has("name","marko").or().where(in("knows").has("name","marko"))
+      """
+    When iterated to list
+    Then the result should be unordered
+      | result |
+      | v[marko] |
+      | v[vadas] |
+      | v[josh]  |
+
+#  Scenario: g_withoutStrategiesXConnectiveStrategyX_V_hasXname_markoX_or_whereXinXknowsX_hasXname_markoXX
```
[the remark this answers is not in the document]
Comment (Cole-Greer) [reply gremlin-test/src/main/resources/org/apache/tinkerpop/gremlin/test/features/integrated/ConnectiveStrategy.feature:35]: Thanks for catching that, I had intended to followup on that but had forgotten. For most of these tests, I used the same traversal for both the `withStrategies` and `withoutStrategies` scenarios. This one is an exception where removal of ConnectiveStrategy needs to change the expected output. The expected output without the strategy active for this traversal is an empty result set. I will update the assertion and re-enable this scenario.

**Pull request merged**